base/src/commonMain/kotlin/utils/timer/MonotonicTimer.kt

```diff
@@ -0,0 +1,135 @@
+/*
+ Copyright 2021 Splendo Consulting B.V. The Netherlands
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+ */
+package com.splendo.kaluga.base.utils.timer
+
+import com.splendo.kaluga.state.HotStateFlowRepo
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.yield
+import kotlin.coroutines.CoroutineContext
+import kotlin.time.Duration
+import kotlin.time.ExperimentalTime
+import kotlin.time.TimeSource
+import com.splendo.kaluga.state.State as KalugaState
+
+@ExperimentalTime
+/** Timer based on the system clock. */
+internal class MonotonicTimer(
+    duration: Duration,
+    coroutineContext: CoroutineContext = Dispatchers.Main.immediate
+) : Timer, HotStateFlowRepo<MonotonicTimer.State>(
+    coroutineContext,
+    {
+        State.NotRunning.Paused(elapsedSoFar = Duration.ZERO, totalDuration = duration)
+    }
+) {
+
+    override val state: StateFlow<Timer.State> = stateFlow
+    override val duration: Duration get() = stateFlow.value.totalDuration
+
+    override suspend fun start() {
+        takeAndChangeState { state ->
+            when (state) {
+                is State.NotRunning.Paused -> suspend {
+                    val finishJob = launch {
```

Comment:
is there a better way to move from `Running` to `Finished`?

my intention was to allow consumer to handle elapsed time update at their own speed so the cold flow was used (see `tickProvider`). that creates a problem of transferring to a `Finished` state as the values might not be read out